### PR TITLE
add new processing algorithms for mesh layer

### PR DESCRIPTION
**Pull request by vcloarec**
New algorithms for mesh layer, ported from crayfish:
- rasterize mesh layer
- export contour to vector layers
- export cross section to text file (csv)
- export time series to text file (csv)

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmexportmesh.cpp [hunk 1]
```diff
@@ -668,4 +694,1021 @@ QSet<int> QgsExportMeshOnGrid::supportedDataType()
...
+
+QString QgsMeshRasterizeAlgorithm::groupId() const
+{
+  return QStringLiteral( "mesh" );
+}
+
+QString QgsMeshRasterizeAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Create a raster layer from a mesh dataset.\n"
+                      "For data on volume (3D stacked dataset values), the exported dataset values are averaged on faces using the method defined in the mesh layer properties (default is Multi level averaging method).\n"
```
Comment (PeterPetrik) [inline src/analysis/processing/qgsalgorithmexportmesh.cpp:720]: I am thinking to NOT allow 3d values in these algorithms at all. And create one extra algorithm "average 3d results", where you would select averaging method and persist it to the new dataset. that way user will be able to first average it out and then export ...

File: src/analysis/mesh/qgsmeshcontours.h [hunk 2]
```diff
@@ -57,21 +57,53 @@ class ANALYSIS_EXPORT QgsMeshContours
...
      */
     QgsMeshContours( QgsMeshLayer *layer );
+
+    /**
+     * Constructs the mesh contours exporter directly with triangular mesh and dataset values to populate the cache with them.
+     * The cache can't be changed.
+     * This instance can be used in a thread safe way.
+     * \param triangularMesh triangular mesh of the meshlayer
+     * \param nativeMesh the native mesh of the mesh layer
+     * \param datasetValues the dataset values to used for creating contours
```
Comment (PeterPetrik) [inline src/analysis/mesh/qgsmeshcontours.h:67]: add they are defined on vertices

File: src/analysis/mesh/qgsmeshcontours.h [hunk 3]
```diff
@@ -57,21 +57,53 @@ class ANALYSIS_EXPORT QgsMeshContours
...
+     * Exports multi line string containing the contour line for particular dataset and value.
+     * \param index dataset index that is used to update the cache
      * \param value value of the contour line
      * \param method for datasets defined on faces, the method will be used to convert data to vertices
      * \param feedback optional feedback object for progress and cancellation support
      * \returns MultiLineString geometry containing contour lines
+     *
+     * \note not thread safe
      */
     QgsGeometry exportLines( const QgsMeshDatasetIndex &index,
```
Comment (PeterPetrik) [inline src/analysis/mesh/qgsmeshcontours.h:90]: I think all these not thread safe methods should be marked as deprecated

File: src/analysis/mesh/qgsmeshcontours.h [hunk 4]
```diff
@@ -57,21 +57,53 @@ class ANALYSIS_EXPORT QgsMeshContours
      * \param layer mesh layer to be associated with this exporter
      */
     QgsMeshContours( QgsMeshLayer *layer );
```
Comment (PeterPetrik) [inline src/analysis/mesh/qgsmeshcontours.h:59]: deprecate?

File: src/analysis/mesh/qgsmeshcontours.h (hunk 3 above)
Comment (vcloarec) [reply src/analysis/mesh/qgsmeshcontours.h:90]: new method are not exposed to python

File: src/analysis/mesh/qgsmeshcontours.h (hunk 4 above)
Comment (vcloarec) [reply src/analysis/mesh/qgsmeshcontours.h:59]: needed for python side

**Pull request merged**